src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspaceDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+internal abstract class AbstractWorkspaceDiagnosticSourceProvider(string name) : IDiagnosticSourceProvider
```

Comment:
not sure how i feel about abstract types here.  For example, i woudl not want disparate features (TODO/EnC) subclassing these to get functionality.  I would be ok with them implementing the interfaces fully, and then having helper static types with utilities they could benefit from (like GetProjectsInPriorityOrder).